[KEP-3521] Part 1: New Pod API .spec.schedulingGates

Huang-Wei opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature
/sig scheduling

#### What this PR does / why we need it:

This is the first part of implementing KEP 3521, including:

- [x] Pod API definition, feature gate, codegen, validation, and drop disabled fields
- [x] disallow binding a Pod carrying non-nil schedulingGates
- [x] disallow creating a Pod with non-nil nodeName and non-nil schedulingGates
- [x] if needed, adds a `{type:PodScheduled, reason:SchedulingGated}` condition upon pod creation
- [x] new literal `SchedulingGated` in the STATUS column of `k get pod`

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of #113269

#### Special notes for your reviewer:

Get best review UX commit by commit.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
New Pod API field `.spec.schedulingGates` is introduced to enable users to control when to mark a Pod as scheduling ready.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3521-pod-scheduling-readiness
```

Huang-Wei commented:
cc @ahg-g 

There are 2 PRs upcoming: scheduler implementation and integration/e2e tests.

ahg-g commented on pkg/apis/core/types.go:
```diff
@@ -3017,6 +3030,13 @@ type PodOS struct {
 }
 
+// PodSchedulingGate is associated to a Pod to guard its scheduling.
+type PodSchedulingGate struct {
```
do we need a struct?

smarterclayton replied:
Yeah, I asked for it for future proofing.

ahg-g commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3272,6 +3272,22 @@ func validateReadinessGates(readinessGates []core.PodReadinessGate, fldPath *fie
 
+func validateSchedulingGates(schedulingGates []core.PodSchedulingGate, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	// There should be no duplicates in list of scheduling gates.
```
```suggestion
	// There should be no duplicates in the list of scheduling gates.
```

ahg-g commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4287,6 +4326,11 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
 	if len(pod.Spec.EphemeralContainers) > 0 {
 		allErrs = append(allErrs, field.Forbidden(fldPath.Child("ephemeralContainers"), "cannot be set on create"))
 	}
+	// A Pod cannot carry non-nil .spec.schedulingGates.
```
```suggestion
	// A Pod cannot be assigned a node if there are remaining scheduling gates.
```

ahg-g commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -642,6 +645,29 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
+		Type:    api.PodScheduled,
+		Status:  api.ConditionFalse,
+		Reason:  api.PodReasonWaitingForGates,
+		Message: "Scheduling is suspended due to non-empty .spec.schedulingGates",
```
```suggestion
		Message: "Scheduling is blocked due to non-empty .spec.schedulingGates",
```

ahg-g commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -642,6 +645,29 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
 
+// applySchedulingWaitingCondition adds a {type:PodScheduled, reason:WaitingForGates} condition
+// to a new-created Pod if necessary.
+func applySchedulingWaitingCondition(pod *api.Pod) {
```
```suggestion
func applyWaitingForSchedulingGatesCondition(pod *api.Pod) {
```

ahg-g commented on pkg/printers/internalversion/printers.go:
```diff
@@ -81,6 +81,9 @@ const (
 	// nodeLabelRole specifies the role of a node
 	nodeLabelRole = "kubernetes.io/role"
+
+	// schedulingPaused indicates a Pod is gated by one or more scheduling readiness gates.
```
I would use blocked instead of pause.

smarterclayton replied:
Agree.  Or SchedulingGated (it should be rare enough that calling it gated will help people find the field.

Huang-Wei replied:
Updated to `schedulingGated` which is more close to the new field name.

ahg-g commented on pkg/printers/internalversion/printers.go:
```diff
@@ -767,6 +770,13 @@ func printPod(pod *api.Pod, options printers.GenerateOptions) ([]metav1.TableRow
+	// If the Pod carries {type:PodScheduled, reason:WaitingForGates}, set reason to SchedulingPaused.
+	for _, condition := range pod.Status.Conditions {
+		if condition.Type == api.PodScheduled && condition.Reason == api.PodReasonWaitingForGates {
+			reason = schedulingPaused
```
why not use the same reason?

Huang-Wei replied:
It's b/c this "reason" is used in the output of `k get pod`, and thus`PodReasonWaitingForGates` which holds the literal "WaitingForGates" loses the context - a word describing that it's scheduling gated would be more self-descriptive.

ahg-g commented:
/assign @smarterclayton

ahg-g commented:
/label api-review

smarterclayton commented on pkg/registry/core/pod/storage/storage.go:
```diff
@@ -221,6 +223,10 @@ func (r *BindingREST) setPodHostAndAnnotations(ctx context.Context, podUID types
 			return nil, fmt.Errorf("pod %v is already assigned to node %q", pod.Name, pod.Spec.NodeName)
 		}
+		// Reject binding to a scheduling un-ready Pod.
+		if utilfeature.DefaultFeatureGate.Enabled(features.PodSchedulingReadiness) && len(pod.Spec.SchedulingGates) != 0 {
```
This should really be in validation - a regular user can patch a pod and set spec.nodeName without binding.

Huang-Wei replied:
ah, update's validation has already been covered. Here it's an additional validation for pod creation - prevent user creating a pod directly with nodeName and non-empty scheduling gates.

Huang-Wei replied:
not quite. L223~L225 is checking whether it's updating nodeName X to Y.

kerthcet replied:
> prevent user creating a pod directly with nodeName and non-empty scheduling gates

I see, misunderstood the words here. Great catch.

smarterclayton commented on pkg/apis/core/types.go:
```diff
@@ -2997,6 +2997,16 @@ type PodSpec struct {
 	// - spec.containers[*].securityContext.runAsGroup
 	// +optional
 	OS *PodOS
+	// Each scheduling gate represents a particular scenario the scheduling is blocked upon.
```
Probably suggest “… scenario that must be satisfied before scheduling.”

smarterclayton commented on pkg/apis/core/types.go:
```diff
@@ -2997,6 +2997,16 @@ type PodSpec struct {
+	// Scheduling is triggered only when SchedulingGates is empty.
+	// In the future, we may impose permission mechanics to restrict which controller can mutate
+	// which scheduling gate. It's dependent on a yet-to-be-implemented fined-grained
+	// permission (https://docs.google.com/document/d/11g9nnoRFcOoeNJDUGAWjlKthowEVM3YGrJA3gLzhpf4)
```
I don’t know that this comment is necessary in the api doc.

smarterclayton commented on pkg/apis/core/types.go:
```diff
@@ -2997,6 +2997,16 @@ type PodSpec struct {
+	// permission (https://docs.google.com/document/d/11g9nnoRFcOoeNJDUGAWjlKthowEVM3YGrJA3gLzhpf4)
+	// and needs to be consistent with how finalizes/liens work today.
+	//
+	// This is an alpha-leve feature enabled by feature gate PodSchedulingReadiness feature gate.
```
typo: “level”

smarterclayton replied:
Also, you only need to say “feature gate” once

Huang-Wei replied:
done. thanks.

smarterclayton commented on pkg/apis/core/types.go:
```diff
@@ -2997,6 +3000,16 @@ type PodSpec struct {
 	// - spec.containers[*].securityContext.runAsGroup
 	// +optional
 	OS *PodOS
+	// Each scheduling gate represents a particular scenario the scheduling is blocked upon.
+	// Scheduling is triggered only when SchedulingGates is empty.
+	// In the future, we may impose permission mechanics to restrict which controller can mutate
+	// which scheduling gate. It's dependent on a yet-to-be-implemented fined-grained
+	// permission (https://docs.google.com/document/d/11g9nnoRFcOoeNJDUGAWjlKthowEVM3YGrJA3gLzhpf4)
+	// and needs to be consistent with how finalizes/liens work today.
```
Also, remember you need to use the camelCase form of field names in public api doc

Huang-Wei replied:
Removed.

smarterclayton commented:
API is pretty close

Huang-Wei commented on pkg/apis/core/types.go:
```diff
@@ -2428,6 +2428,9 @@ const (
 	// can't schedule the pod right now, for example due to insufficient resources in the cluster.
 	PodReasonUnschedulable = "Unschedulable"
+	// PodReasonWaitingForGates reason in PodScheduled PodCondition means that the scheduler
+	// suspends scheduling the pod for now, because it carries non-nil .spec.schedulingGates.
```
reworded a bit:

```
// skips scheduling the pod because one or more scheduling gates are still present.
```

Huang-Wei commented:
@ahg-g @smarterclayton thanks for the review during kubecon! All comments have been addressed. PTAL.

leilajal commented:
/remove-sig api-machinery

ahg-g commented on pkg/apis/core/types.go:
```diff
@@ -3000,14 +3000,12 @@ type PodSpec struct {
-	// which scheduling gate. It's dependent on a yet-to-be-implemented fined-grained
-	// permission (https://docs.google.com/document/d/11g9nnoRFcOoeNJDUGAWjlKthowEVM3YGrJA3gLzhpf4)
-	// and needs to be consistent with how finalizes/liens work today.
+	// which scheduling gate.
```
The comment regarding the future is still there, I don't think this is useful in the api. Other fields point to the KEP, so why not just say: 

```
// More info: https://git.k8s.io/enhancements/keps/sig-scheduling/3521-pod-scheduling-readiness
```

ahg-g commented on pkg/apis/core/types.go:
```diff
@@ -3000,14 +3000,12 @@ type PodSpec struct {
 	OS *PodOS
-	// Each scheduling gate represents a particular scenario the scheduling is blocked upon.
-	// Scheduling is triggered only when SchedulingGates is empty.
+	// Each scheduling gate represents a particular scenario that must be satisfied before scheduling.
```
Nit: can we be consistent with the description we have for finalizers?

"// schedulingGates is an opaque list of values that if specified will block scheduling the pod."

ahg-g reviewed (commented):
nits

Huang-Wei commented:
/retest

ahg-g commented:
/lgtm

Huang-Wei committed:
APIs, Validation and condition enforcements

- New API field .spec.schedulingGates
- Validation and drop disabled fields
- Disallow binding a Pod carrying non-nil schedulingGates
- Disallow creating a Pod with non-nil nodeName and non-nil schedulingGates
- Adds a {type:PodScheduled, reason:WaitingForGates} condition if necessary
- New literal SchedulingGated in the STATUS column of `k get pod`

Huang-Wei committed:
Automated codegen

Huang-Wei commented:
Thanks @ahg-g. Squshed into two commits.

@soltysh would you (or delegate to someone) please review the CLI changes? In particular, the status volume of `k get pod` will output "SchedulingGated" if the pod carries non-empty .spec.schedulingGates.

Huang-Wei commented:
/retest

Huang-Wei commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3272,6 +3272,22 @@ func validateReadinessGates(readinessGates []core.PodReadinessGate, fldPath *fie
+	// There should be no duplicates in the list of scheduling gates.
+	seen := sets.String{}
+	for i, schedulingGate := range schedulingGates {
+		if schedulingGate.Name == "" {
```
Finalizers don't enforce it but the network readiness gate does. So it seems there is no clear pattern to enforce qualified name or not, I'd prefer relax the naming here.

smarterclayton replied:
I think we should consider it - but can be a follow up.  Can you spawn an issue?

Huang-Wei commented on pkg/apis/core/types.go:
```diff
@@ -2428,6 +2428,9 @@ const (
 	PodReasonUnschedulable = "Unschedulable"
+	// PodReasonWaitingForGates reason in PodScheduled PodCondition means that the scheduler
+	// skips scheduling the pod because one or more scheduling gates are still present.
+	PodReasonWaitingForGates = "WaitingForGates"
```
hrm, this reason literal is applied to the PodScheduled status, so it'd look like this:

```yaml
- type: PodScheduled
  status: False
  reason: WaitingForGates
  message: Scheduling is blocked due to non-empty scheduling gates
```

IMO it looks descriptive in the whole context, or do you think it needs some polish?

ahg-g replied:
I think `SchedulingGated` makes sense, and we can use it in the kubectl output as well as is.

Huang-Wei replied:
no strong option. updated.

ahg-g commented on pkg/apis/core/types.go:
```diff
@@ -2428,9 +2428,9 @@ const (
+	// PodReasonSchedulingGated reason in PodScheduled PodCondition means that the scheduler
 	// skips scheduling the pod because one or more scheduling gates are still present.
-	PodReasonWaitingForGates = "WaitingForGates"
+	PodReasonSchedulingGated = "WaitingForGates"
```
change the constant value :)

Huang-Wei replied:
thought I did it...

ahg-g commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2666,9 +2666,9 @@ const (
+	// PodReasonSchedulingGated reason in PodScheduled PodCondition means that the scheduler
 	// skips scheduling the pod because one or more scheduling gates are still present.
-	PodReasonWaitingForGates = "WaitingForGates"
+	PodReasonSchedulingGated = "WaitingForGates"
```
here too

ahg-g commented:
/lgtm

Great! @smarterclayton I think this is ready. Note that the implementation PR is also ready to merge once this one does https://github.com/kubernetes/kubernetes/pull/113275

fedebongio commented:
/remove-sig api-machinery

smarterclayton commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3327,6 +3331,16 @@ type PodSpec struct {
 	// +optional
 	HostUsers *bool `json:"hostUsers,omitempty" protobuf:"bytes,37,opt,name=hostUsers"`
+	// SchedulingGates is an opaque list of values that if specified will block scheduling the pod.
+	// More info:  https://git.k8s.io/enhancements/keps/sig-scheduling/3521-pod-scheduling-readiness.
```
I'm going to approve this as is but I want to add a better description before we hit beta.

smarterclayton commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4287,6 +4326,11 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
+	// A Pod cannot be assigned a Node if there are remaining scheduling gates.
+	if utilfeature.DefaultFeatureGate.Enabled(features.PodSchedulingReadiness) &&
+		pod.Spec.NodeName != "" && len(pod.Spec.SchedulingGates) != 0 {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("nodeName"), "cannot be set along with non-nil schedulingGates"))
```
Don't use `non-nil` - suggest `cannot be set until all schedulingGates have been cleared`

Huang-Wei replied:
updated.

smarterclayton commented:
If you can change the validation string i'll approve - everything else can be improved before beta.

Huang-Wei commented:
> If you can change the validation string i'll approve - everything else can be improved before beta.

Thanks. Tracked in https://github.com/kubernetes/kubernetes/issues/113608

smarterclayton commented:
/approve

ahg-g commented:
/lgtm

Huang-Wei commented:
/hold
for CI.

Huang-Wei commented:
/unhold

ahg-g commented:
/hold cancel

fedebongio commented:
/remove-sig api-machinery

brandond commented:
@Huang-Wei @smarterclayton there is a typo the PR comment that made it into the CHANGELOG.

From the user-facing change section (that is scraped by the changelog generator), `.spec.scheudlingGates` should be `.spec.schedulingGates`

Huang-Wei commented:
@brandond thanks for the catch.. Rasied a PR fixing the typos: #113843
